torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -22,73 +22,95 @@ AccumulateGrad::AccumulateGrad(Variable variable_)
-      // changing the tensor itself, and thus in this case we have to change the grad
-      // tensor.
-      grad_variable = new_grad + grad_variable;
+    if (variable_grad.is_sparse() && !new_grad_copy.is_sparse()) {
+      // If `grad_variable` is sparse and `new_grad_copy` is not sparse, their
+      // sum is not sparse, and we must change the TensorImpl type of
+      // `grad_variable` for it to store the result. However, changing the
+      // TensorImpl type of a tensor requires changing the tensor itself, and
+      // thus in this case we have to change the grad tensor.
+      update_grad(new_grad_copy + variable_grad);
     } else {
-      // In this case we can avoid changing the grad tensor. There are three scenarios
-      // when we'll hit this case:
+      // In this case we can avoid changing the grad tensor. There are three
+      // scenarios when we'll hit this case:
       //
-      // 1. `grad_variable` is sparse, and `new_grad` is sparse.
-      // 2. `grad_variable` is dense, and `new_grad` is sparse.
-      // 3. `grad_variable` is dense, and `new_grad` is dense.
+      // 1. `grad_variable` is sparse, and `new_grad_copy` is sparse.
+      // 2. `grad_variable` is dense, and `new_grad_copy` is sparse.
+      // 3. `grad_variable` is dense, and `new_grad_copy` is dense.
       //
-      // In all of these three cases, `grad_variable += new_grad` is a valid operation
-      // which adds `new_grad` to `grad_variable` in place. `grad_variable` is thus
-      // still referring to the same tensor after the operation.
-      grad_variable += new_grad;
+      // In all of these three cases, `grad_variable += new_grad_copy` is a
+      // valid operation which adds `new_grad_copy` to `grad_variable` in place.
+      // `grad_variable` is thus still referring to the same tensor after the
+      // operation.
+      variable_grad += new_grad_copy;
     }
   } else {
-    variable.grad() = grad + new_grad;
+    update_grad(variable_grad + new_grad_copy);
   }
+}
+
+auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
+  // XXX: this method is not thread-safe!
+  check_input_variables("AccumulateGrad", grads, 1, 0);
+
+  if (!grads[0].defined())
+    return {};
+  if (variable.grad_fn())
+    throw std::logic_error(
+        "leaf variable has been moved into the graph interior");
+  if (!variable.requires_grad())
+    return {};
+
+  const auto& new_grad = grads[0];
+  at::Tensor& grad = variable.grad();
+  // If the function has post hooks (for example, a DDP allreduce hook),
+  // call_function in Engine.cpp will temporarily bump the refcount by one,
+  // hence the addition of !post_hooks().empty() for num_expected_refs in
+  // addition to the one reference that we're holding.
+  accumulateGradAndCallHooks(
+      variable,
+      grad,
+      new_grad,
+      1 + !post_hooks().empty() /* num_expected_refs */,
+      [&grad](at::Tensor&& grad_update) { grad = grad_update; });
```

Comment:
If you're going to take `grad_update` by rvalue reference, you should cash in on it and `grad = std::move(grad_update)`